File: asp/server/operator-service.ts
/** Desktop operator commands and conversation views over the session service. */

import { randomBytes } from "node:crypto";

import { NotAllowed, NotFound, SessionService } from "./service.ts";
import { Store } from "./store.ts";
import type { Transport } from "./transport.ts";

export class OperatorService {
  readonly sessions: SessionService;
  readonly store: Store;
  readonly transport: Transport;

  constructor(sessions: SessionService, store: Store, transport: Transport) {
    this.sessions = sessions;
    this.store = store;
    this.transport = transport;
  }

  async createConversation(creator: string, kind: string): Promise<Record<string, unknown>> {
    if (kind !== "bot" && kind !== "group") {
      throw new Error("kind must be bot or group");
    }
    let invite: string[] = [];
    let magi: { handle: string; name: string; token: string } | null = null;
    if (kind === "bot") {
      const magiName = this.store.nextMagiName();
      const handle = Store.magiHandle(magiName);
      const magiToken = randomBytes(24).toString("base64url");
      this.store.registerAgent({ handle, token: magiToken, name: magiName, managed: true });
      magi = { handle, name: magiName, token: magiToken };
      invite = [handle];
    }
    const result = await this.sessions.createSession({
      creator,
      invite,
      topic: null,
      initialMessage: null,
      endAfterSend: false,
    });
    const session = this.store.getSession(result.sessionId);
    if (session !== undefined) {
      session.kind = kind;
      this.store.updateSession(session);
    }
    const view = this.conversationView(creator, result.sessionId);
    if (magi !== null) {
      view.name = magi.name;
      // ASP registers the agent but never starts it: the caller (the desktop
      // app on this machine) runs the process itself, and this is where it
      // gets the handle and token to start it with.
      view.magi = { handle: magi.handle, name: magi.name, token: magi.token };
    }
    return view;
  }

  /**
   * Every registered agent with the credential its runner needs. Operator only:
   * a MAGI token authenticates a live socket, not the operator API.
   */
  listAgents(): Record<string, unknown>[] {
    const agents: Record<string, unknown>[] = [];
    for (const [handle, agent] of this.store.agents) {
      agents.push({
        handle,
        token: agent.token,
        name: agent.name,
        nickname: agent.nickname,
        managed: agent.managed,
        online: this.transport.isOnline(handle),
      });
    }
    return agents;
  }

  conversationView(caller: string, sessionId: string): Record<string, unknown> {
    const view = this.sessions.getSessionView(caller, sessionId);
    const session = this.store.getSession(sessionId);
    const agents = this.store
      .participantsIn(sessionId)
      .filter(
        (participant) =>
          participant.handle !== caller &&
          (participant.status === "invited" || participant.status === "joined"),
      )
      .map((participant) => participant.handle);
    view.conversation_id = sessionId;
    view.agents = agents;
    if (session !== undefined) {
      view.kind = session.kind ?? (agents.length === 1 ? "bot" : "group");
      if (session.description !== null) {
        view.description = session.description;
      }
    }
    if (view.kind === "bot" && agents.length === 1) {
      const agent = this.store.getAgent(agents[0] ?? "");
      const handle = agents[0];
      if (agent !== undefined && handle !== undefined) {
        view.name = agent.nickname ?? agent.name ?? handle;
      }
    }
    return view;
  }

  listConversations(caller: string): Record<string, unknown>[] {
    const rows: Record<string, unknown>[] = [];
    for (const session of this.store.sessions.values()) {
      if (this.store.getParticipant(session.id, caller) === undefined) {
        continue;
      }
      rows.push(this.conversationView(caller, session.id));
    }
    rows.sort((left, right) => Number(right.created_at ?? 0) - Number(left.created_at ?? 0));
    return rows;
  }

  listBots(caller: string, conversationId: string | null = null): Record<string, unknown>[] {
    let inConversation: Set<string> | null = null;
    if (conversationId !== null) {
      if (
        this.store.getSession(conversationId) === undefined ||
        this.store.getParticipant(conversationId, caller) === undefined
      ) {
        throw new NotFound();
      }
      inConversation = new Set(
        this.store
          .participantsIn(conversationId)
          .filter(
            (participant) => participant.status === "invited" || participant.status === "joined",
          )
          .map((participant) => participant.handle),
      );
    }
    const bots: Record<string, unknown>[] = [];
    for (const handle of this.store.agents.keys()) {
      if (handle === caller) {
        continue;
      }
      const agent = this.store.getAgent(handle);
      const name = agent === undefined ? null : agent.nickname ?? agent.name;
      const row: Record<string, unknown> = {
        handle,
        name: name ?? handle,
        online: this.transport.isOnline(handle),
      };
      if (inConversation !== null) {
        row.in_conversation = inConversation.has(handle);
      }
      bots.push(row);
    }
    bots.sort((left, right) => {
      const a = String(left.handle);
      const b = String(right.handle);
      return a < b ? -1 : a > b ? 1 : 0;
    });
    return bots;
  }

  async addConversationMember(
    caller: string,
    sessionId: string,
    handle: string,
  ): Promise<Record<string, unknown>> {
    if (handle === caller) {
      throw new NotAllowed();
    }
    if (this.store.getAgent(handle) === undefined) {
      throw new NotFound();
    }
    await this.sessions.invite(caller, sessionId, [handle]);
    return this.conversationView(caller, sessionId);
  }

  async updateConversation(
    caller: string,
    sessionId: string,
    topic: string | null,
    description: string | null,
  ): Promise<Record<string, unknown>> {
    await this.sessions.updateSession(caller, sessionId, topic, description);
    return this.conversationView(caller, sessionId);
  }
}

File: magi-asp/server/operator_service.py
"""Desktop operator commands and conversation views over the session service."""

from __future__ import annotations

import secrets
from typing import Any

from .service import NotAllowed, NotFound, SessionService
from .spawn import MagiSpawner, SpawnedMagi, spawn_to_wire
from .store import Store
from .transport import Transport


class OperatorService:
    def __init__(
        self,
        sessions: SessionService,
        store: Store,
        transport: Transport,
        spawner: MagiSpawner,
        base_url: str,
    ) -> None:
        self.sessions = sessions
        self.store = store
        self.transport = transport
        self.spawner = spawner
        self.base_url = base_url

    async def create_conversation(self, creator: str, kind: str) -> dict[str, Any]:
        """Operator create: ASP spawns a MAGI for `bot`, or opens an empty group.

        `kind` is the plus-button action, not a lasting conversation type.
        The desktop only POSTs this; it must not start MAGI itself.
        ASP assigns MAGI `name` values eva-000, eva-001, …
        """
        if kind not in ("bot", "group"):
            raise ValueError("kind must be bot or group")
        invite: list[str] = []
        spawned: SpawnedMagi | None = None
        magi_token: str | None = None
        magi_name: str | None = None
        if kind == "bot":
            magi_name = self.store.next_magi_name()
            handle = self.store.magi_handle(magi_name)
            magi_token = secrets.token_urlsafe(24)
            self.store.register_agent(handle, magi_token, name=magi_name, managed=True)
            spawned = self.spawner.spawn(handle=handle, base=self.base_url, token=magi_token)
            invite = [handle]
        result = await self.sessions.create_session(
            creator=creator,
            invite=invite,
            topic=None,
            initial_message=None,
            end_after_send=False,
        )
        sess = self.store.get_session(result.session_id)
        if sess is not None:
            sess.kind = kind
            self.store.update_session(sess)
        view = self.conversation_view(creator, result.session_id)
        view["spawned"] = bool(spawned and spawned.spawned)
        if magi_name is not None:
            view["name"] = magi_name
        if spawned is not None:
            wire = dict(spawn_to_wire(spawned))
            if magi_token is not None:
                wire["token"] = magi_token
            if magi_name is not None:
                wire["name"] = magi_name
            view["magi"] = wire
        return view

    def conversation_view(self, caller: str, session_id: str) -> dict[str, Any]:
        view = self.sessions.get_session_view(caller, session_id)
        sess = self.store.get_session(session_id)
        agents = [
            p.handle
            for p in self.store.participants_in(session_id)
            if p.handle != caller and p.status in ("invited", "joined")
        ]
        view["conversation_id"] = session_id
        view["agents"] = agents
        if sess is not None:
            view["kind"] = sess.kind or ("bot" if len(agents) == 1 else "group")
            if sess.description is not None:
                view["description"] = sess.description
        if view.get("kind") == "bot" and len(agents) == 1:
            agent = self.store.get_agent(agents[0])
            if agent is not None:
                view["name"] = agent.nickname or agent.name or agents[0]
        return view

    def list_conversations(self, caller: str) -> list[dict[str, Any]]:
        out: list[dict[str, Any]] = []
        for sess in self.store.sessions.values():
            if self.store.get_participant(sess.id, caller) is None:
                continue
            out.append(self.conversation_view(caller, sess.id))
        out.sort(key=lambda row: row.get("created_at") or 0, reverse=True)
        return out

    def list_bots(
        self, caller: str, conversation_id: str | None = None
    ) -> list[dict[str, Any]]:
        """MAGI this operator can add — registered agents, not a static roster."""
        in_conversation: set[str] = set()
        if conversation_id is not None:
            if self.store.get_session(conversation_id) is None:
                raise NotFound()
            if self.store.get_participant(conversation_id, caller) is None:
                raise NotFound()
            in_conversation = {
                participant.handle
                for participant in self.store.participants_in(conversation_id)
                if participant.status in ("invited", "joined")
            }
        bots: list[dict[str, Any]] = []
        for handle in self.store.agents:
            if handle == caller:
                continue
            agent = self.store.get_agent(handle)
            name = (agent.nickname or agent.name) if agent is not None else None
            row: dict[str, Any] = {
                "handle": handle,
                "name": name or handle,
                "online": self.transport.is_online(handle),
            }
            if conversation_id is not None:
                row["in_conversation"] = handle in in_conversation
            bots.append(row)
        bots.sort(key=lambda row: row["handle"])
        return bots

    async def add_conversation_member(
        self, caller: str, session_id: str, handle: str
    ) -> dict[str, Any]:
        if handle == caller:
            raise NotAllowed()
        if self.store.get_agent(handle) is None:
            raise NotFound()
        await self.sessions.invite(caller, session_id, [handle])
        return self.conversation_view(caller, session_id)

    async def update_conversation(
        self, caller: str, session_id: str, topic: str | None, description: str | None
    ) -> dict[str, Any]:
        await self.sessions.update_session(caller, session_id, topic, description)
        return self.conversation_view(caller, session_id)
